one/api.py

```diff
@@ -1593,67 +1623,101 @@ def _tag_mismatched_file_record(self, url):
+        if isinstance(url, list):
+            # download all files in the list of urls, store file paths with UUID to variable
+            local_path_list = self.alyx.download_file(url, cache_dir=str(target_dir))
+            # check hash and file_size mismatches on downloaded files
+            if isinstance(hash, list) | isinstance(file_size, list):
+                for single_file_size, single_hash, local_path, single_url in \
+                        zip(file_size, hash, local_path_list, url):
+                    self._check_hash_and_file_size_mismatch(single_file_size, single_hash,
+                                                            local_path, target_dir, single_url)
+            # check if we are keeping the uuid on the list of file names
+            if keep_uuid:
+                return local_path_list
+            else:  # remove uuids from list of file names
+                local_path_list_to_return = []
+                for local_path in local_path_list:
+                    local_path_list_to_return.append(alfio.remove_uuid_file(local_path))
+                return local_path_list_to_return
+
+        else:  # url is not a list
+            if not offline:
+                local_path = self.alyx.download_file(url, cache_dir=str(target_dir))
+                if isinstance(file_size, int) | isinstance(hash, str):
+                    self._check_hash_and_file_size_mismatch(file_size, hash, local_path,
+                                                            target_dir, url)
+
+                # check if we are keeping the uuid on the list of file names
+                if keep_uuid:
+                    return local_path
+                else:  # remove uuids from filenames
+                    return alfio.remove_uuid_file(local_path)
+
+    def _check_hash_and_file_size_mismatch(self, file_size, hash, local_path, target_dir, url):
+        """
+        Check to ensure the hash and file size of a downloaded file matches what is on disk
+
+        Parameters
+        ----------
+        file_size : int
+            The expected file size to compare with downloaded file
+        hash : str
+            The expected file hash to compare with downloaded file
+        local_path: str
+            The path of the downloaded file
+        target_dir : str
+            The root directory of the downloaded file
+        url : str
+            An absolute or relative URL for a remote dataset
+        """
+        # verify has size
+        hash_mismatch = hash and hashfile.md5(Path(local_path)) != hash
+        # verify file size
+        file_size_mismatch = file_size and Path(local_path).stat().st_size != file_size
+        # check if there is a mismatch in hash or file_size
+        if hash_mismatch or file_size_mismatch:
+            _logger.warning(f'local md5 or size mismatch, re-downloading {local_path}')
+            local_path, md5 = self.alyx.download_file(url, cache_dir=str(target_dir),
```

Comment:
It doesn't make sense to re-download the file if the hash doesn't match.  I'll fix up this branch then merge it

Reply:
Ok, sounds good to me.